internal/grpc/grpclog/grpclog_test.go

```diff
@@ -0,0 +1,102 @@
+/*
+Copyright (c) Edgeless Systems GmbH
+
+SPDX-License-Identifier: AGPL-3.0-only
+*/
+
+// grpclog provides a logging utilities for gRPC.
+package grpclog
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"google.golang.org/grpc/connectivity"
+)
+
+func TestLogStateChanges(t *testing.T) {
+	testCases := []struct {
+		name    string
+		conn    getStater
+		isReady bool
+		assert  func(t *testing.T, lg *fakeLog, isReady bool)
+	}{
+		{
+			name: "log state changes",
+			conn: &fakeConn{
+				states: []connectivity.State{
+					connectivity.Connecting,
+					connectivity.Ready,
+				},
+			},
+			isReady: true,
+			assert: func(t *testing.T, lg *fakeLog, isReady bool) {
+				assert.Equal(t, "Connection state started as CONNECTING", lg.msgs[0])
+				assert.Equal(t, "Connection state changed to CONNECTING", lg.msgs[1])
+				assert.Equal(t, "Connection ready", lg.msgs[2])
+				assert.Len(t, lg.msgs, 3)
+			},
+		},
+		{
+			name: "WaitForStateChange returns false (e.g. when context is canceled)",
+			conn: &fakeConn{
+				states: []connectivity.State{
+					connectivity.Connecting,
+					connectivity.Idle,
+				},
+				stopWaitForChange: true,
+			},
+			isReady: false,
+			assert: func(t *testing.T, lg *fakeLog, isReady bool) {
+				assert.Equal(t, "Connection state started as CONNECTING", lg.msgs[0])
+				assert.Equal(t, "Connection state ended with CONNECTING", lg.msgs[1])
+				assert.Len(t, lg.msgs, 2)
+				assert.False(t, isReady)
+			},
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			lg := &fakeLog{}
+
+			var wg sync.WaitGroup
+			isReady := false
+			LogStateChanges(context.Background(), tc.conn, lg, &wg, func() { isReady = true })
+			wg.Wait()
+			tc.assert(t, lg, isReady)
+		})
+	}
+}
+
+type fakeLog struct {
```

Comment:
This is a stub, not a fake.

Reply:
We also have a [logger for unit tests](https://github.com/edgelesssys/constellation/blob/25037026e150873447bafb693f8be9b93e5afe1f/internal/logger/log.go#L98)

Reply:
> 

Yes i saw it, but I didn't manage to properly inspect it for the test. The logger has timestamps which make it impossible to assert its output. Also I found it difficult to redirect the logging output so that I can compare it in the test. Do you know how to fix that @daniel-weisse ? Feel free to call :)

Reply:
Ah sry, didnt read the code parts that actually make use of the logger.
If you actually want to assert that certain messages were written to the logger, this is probably not the correct one to use.

Depending on how/where the `LogStateChanges` is used, would it maybe be easier to pass an `io.Writer` instead of `*logger.Logger`?

Reply:
oh yeah using an io.Writer in the spyLogger could have been an option

Reply:
It returns canned values, so it is a stub. It is also a spy, but the spy pattern is somewhat complementary to the stub/fake/mock classification.